### PR TITLE
Remove tx_history and ledger_header in API version 2

### PR DESCRIPTION
**Pull request by Bronek**
## High Level Overview of Change

Remove `tx_history` and `ledger_header` methods from API version 2. Also update `RPC::Handler` to allow for methods (or methods implementation) to be version specific. This partially resolves https://github.com/XRPLF/rippled/issues/4727 and fully resolves https://github.com/XRPLF/rippled/issues/3638 and https://github.com/XRPLF/rippled/issues/3539 

### Context of Change

Note design change in `RPC::Handler` - we can now store multiple handlers with the same name, as long as they belong to different (non-overlapping) API versions. This necessarily impacts the handler lookup algorithm and its complexity, however there is no performance loss on x86_64 architecture, and only minimal performance loss on arm64 (around 10ns). This design change gives us extra flexibility evolving API in the future, including other parts of https://github.com/XRPLF/rippled/issues/4727 

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Before

In any API version `tx_history` and `ledger_header` are recognised methods

### After
In API version 1 `tx_history` and `ledger_header` are a valid API method
In API version 2 `tx_history` and `ledger_header` are not recognised method; `rippled` will return error `unknownCmd` if they are called.

### CODE REVIEW

**Comment by mDuo13** [comment]
Fixes #3638

**Comment by mDuo13** [comment]
P.S. while we're at it, maybe do #3539 also?

File: API-CHANGELOG.md
Comment (mDuo13) [inline API-CHANGELOG.md:133]: The point of `tx_history` is to get a list of recently applied transactions. The `account_tx` and `ledger` methods are the next best thing for that effect. (I did find one narrow case where tx_history is actually more useful, which is on a chain with very little activity, it can give you recent transactions with fewer steps/foreknowledge than the other two. That need is very niche, mostly served by the ledger command, and less useful because of tx_history's non-standard format, so it's still fine to deprecate IMO.) The `tx` method just looks up an individual transaction if you already know its hash, which is different enough that I think we should not suggest it here.

The `binary` note on `ledger_header` is a red herring; both `ledger` and `ledger_header` return the same (non-binary) info by default. Binary support was the last thing on the `ledger_header` side of the venn diagram, but now the `ledger` command is a superset of the functionality.

```suggestion
- `tx_history` - Instead, use other methods such as `account_tx` or `ledger` with the `transactions` field set to `true`.
- `ledger_header` - Instead, use the `ledger` method.
```

P.S. The word "please" can convey a lot of different tones, not all of them good, especially if we only include it in some places and not others. For the changelog, I think it's best to omit the word "please" from these types of notes.

**Review by mDuo13 — COMMENTED** [review]
Excited to finally get rid of some of this old cruft!

**Comment by Bronek** [comment]
Apologies to reviewers about all these `forced push`, I was trying to make the micro-benchmark `Handler_test.cpp` little less stupid. Which happens to be first commit in the series because I use it to establish baseline ...

File: API-CHANGELOG.md
Comment (Bronek) [reply API-CHANGELOG.md]: Thanks !

File: API-CHANGELOG.md
Comment (Bronek) [reply API-CHANGELOG.md]: Thanks !

**Comment by intelliot** [comment]
@Bronek - please comment here to confirm when this PR is ready for review

**Comment by Bronek** [comment]
This is ready for review, thanks !

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [inline src/ripple/rpc/impl/Handler.cpp:219]: Instead of the `assert`, can we throw a `LogicError` instead? I'm not sure if asserts are enabled in all types of builds. If there are duplicate/overlapping handlers, it sounds like we need to throw a halting error message.

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [inline src/ripple/rpc/impl/Handler.cpp:266]: Can we return a `std::set<char const*>` instead of manually checking for duplicates in the vector?

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [inline src/ripple/rpc/impl/Handler.cpp:283]: At first blush, this check appears to be redundant. We are performing this check  in the `HandlerTable` constructor, do we need to repeat ourselves?

`addHandler` is not being invoked from any other place in the codebase.

File: src/test/rpc/Handler_test.cpp
Comment (ckeshava) [inline src/test/rpc/Handler_test.cpp:98]: thanks for computing the metrics. can you also include the unit of time in the print statements? I believe it's micro-seconds for this line.

**Review by ckeshava — CHANGES_REQUESTED** [review]
thanks Bronek, I've put down some comments

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [reply src/ripple/rpc/impl/Handler.cpp:283]: I believe we can remove the check in the `HandlerTable` constructor. I verified that the existing unit tests work correctly without this check.

But this is a suggestion, please don't feel its binding to follow it.

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [reply src/ripple/rpc/impl/Handler.cpp:266]: I've made the necessary changes for using a `std::set` in at the tip of this [branch](https://github.com/ckeshava/rippled/tree/handlerSet)

feel free use it. this is the [commit hash](b705fc4a3c0ce73e6e93480134e1027ead34ab17)

File: src/test/rpc/TransactionHistory_test.cpp
Comment (Bronek) [reply src/test/rpc/TransactionHistory_test.cpp:69]: it's in newly added  `src/test/rpc/LedgerHeader_test.cpp`

File: src/ripple/rpc/impl/Handler.cpp
Comment (Bronek) [reply src/ripple/rpc/impl/Handler.cpp:219]: sure, can do. Note that debug binary is tested in CI (meaning, an assert would be caught), and the population of handlers is exactly the same in both debug and release.

File: src/ripple/rpc/impl/Handler.cpp
Comment (Bronek) [reply src/ripple/rpc/impl/Handler.cpp:266]: thank you, will grab your commit - agree that's an improvement.

File: src/test/rpc/Handler_test.cpp
Comment (Bronek) [reply src/test/rpc/Handler_test.cpp:98]: will do, thanks !

File: src/test/rpc/Handler_test.cpp
Comment (Bronek) [inline src/test/rpc/Handler_test.cpp:35]: @HowardHinnant would you like to comment on this ?

**Comment by intelliot** [comment]
(note: these API methods weren't in Clio, so no need for Clio to change anything here)

File: src/test/rpc/Handler_test.cpp
Comment (ckeshava) [inline src/test/rpc/Handler_test.cpp:113]: Why do we need to insert this `dummy` computation? Won't it add extra time into the results?

File: src/test/rpc/Handler_test.cpp
Comment (ckeshava) [inline src/test/rpc/Handler_test.cpp:133]: Can you add the extra line at the end of this file? I think it's part of coding convention, it's followed in all the other files.

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [inline src/ripple/rpc/impl/Handler.cpp:26]: `chrono` and `type-traits` includes are not required

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [inline src/ripple/rpc/impl/Handler.cpp:63]: unrelated to your PR, but there is a stray `;` ^^

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [inline src/ripple/rpc/impl/Handler.cpp:265]: I forgot to add a `[[nodiscard]]` tag, can you please add that?

**Review by ckeshava — COMMENTED** [review]
I have a few clarifying questions for this PR.

Although the unit tests are sufficient, I'd like to test the API changes via the command line. How can I enable the API-version 2 in my local `rippled`? I set `beta_rpc_api` to `1` in the config file, but that didn't seem to help.

File: src/ripple/rpc/handlers/Version.h
Comment (ckeshava) [inline src/ripple/rpc/handlers/Version.h:53]: Should we add an assert/check in the [constructor](https://github.com/XRPLF/rippled/blob/1fde585003c286841e3884012bef000d1ff5f4ee/src/ripple/rpc/handlers/Version.h#L33) of `VersionHandler` for the compatibility of `apiVersion_`? 

I'm trying to understand the expected behavior for wrong `apiVersion` inputs. At the moment, the unit tests fail (the `result` is a `null` JSON Value). I feel it would be nice to return a descriptive error message for the end-user (Ex: unknown version ?)


We need to ensure that `VersionHandler` and `LedgerHandler` behave in a similar way.

Comment (ckeshava) [reply src/ripple/rpc/handlers/Version.h:53]: To reproduce the unit test: I set the `apiVersion = 4` [here](https://github.com/Bronek/rippled/blob/6de4b6c175d80880901f7f95c9f05421152adfb7/src/test/rpc/TransactionHistory_test.cpp#L65)

File: src/test/rpc/Handler_test.cpp
Comment (HowardHinnant) [reply src/test/rpc/Handler_test.cpp:35]: MSVC (recent versions) have C++20 chrono.  gcc-14 has it, but hasn't shipped yet.  And LLVM is not close.

File: src/ripple/rpc/handlers/Version.h
Comment (Bronek) [reply src/ripple/rpc/handlers/Version.h:53]: I will add assert in `template < . . . class HandlerImpl> handle(JsonContext& context, Object& object)` which is shared by both `LedgerHandler` and `VersionHandler` ; as for the actual behaviour, we expect the user to get `unknownCmd` as presented in unit  tests, no plan for more diagnostic. We get that for the simplicity of handler lookup , in `getHandler`. We can certainly document that this error could be triggered by invalid selection of api version, but that probably belongs to https://xrpl.org  ( @mDuo13 - what's your opinion ?)

File: src/test/rpc/Handler_test.cpp
Comment (ckeshava) [reply src/test/rpc/Handler_test.cpp:113]: i see, that's an interesting point 👍

File: src/ripple/rpc/impl/Handler.cpp
Comment (ckeshava) [inline src/ripple/rpc/impl/Handler.cpp:54]: My compiler (apple-clang v15) is throwing this error on these two lines:
```
/Users/ckeshavabs/personal-code/rippled/src/ripple/rpc/impl/Handler.cpp:54:31: error: 'HandlerImpl' does not refer to a value
        context.apiVersion >= HandlerImpl.minApiVer &&
                              ^
/Users/ckeshavabs/personal-code/rippled/src/ripple/rpc/impl/Handler.cpp:49:31: note: declared here
template <class Object, class HandlerImpl>
                              ^
/Users/ckeshavabs/personal-code/rippled/src/ripple/rpc/impl/Handler.cpp:55:31: error: 'HandlerImpl' does not refer to a value
        context.apiVersion <= HandlerImpl.maxApiVer);
                              ^
/Users/ckeshavabs/personal-code/rippled/src/ripple/rpc/impl/Handler.cpp:49:31: note: declared here
template <class Object, class HandlerImpl>
```

Did you mean to use the scope-resolution operator? `HandlerImpl::minApiVer` -> this builds for me without a hitch

File: src/ripple/rpc/handlers/Version.h
Comment (ckeshava) [reply src/ripple/rpc/handlers/Version.h:53]: Thank you, I verified that this works correctly.

I believe it is better to not construct incorrect `VersionHandler` and `LedgerHandler` objects. Although the `assert` in the `handle` function catches this error, it'd probably be better to catch this issue in the constructors itself, as early as possible. (or possibly in `RPCHelpers.cpp`, even before the ctors)

**Comment by ckeshava** [comment]
This PR looks good. I have left a few more comments, but they are non-binding suggestions.

I want to draw attention to this earlier question (https://github.com/XRPLF/rippled/pull/4759#pullrequestreview-1683130236)
I compile and run this branch of `rippled` on my local machine. The command `rippled tx_history 0` displays old output, I do not get the expected `unknownCmd` error. Isn't this wrong?
The version is set to `2.0.0-b3`([here](https://github.com/XRPLF/rippled/blob/b421945e717a29e81b029adea576b24cd2068051/src/ripple/protocol/impl/BuildInfo.cpp#L36)) Do I need to do anything else to switch to version 2 for my local rippled server?

The unit tests are satisfactory, but I'm not able to verify this change through a local version of `rippled`.

File: src/ripple/rpc/impl/Handler.cpp
Comment (Bronek) [reply src/ripple/rpc/impl/Handler.cpp]: you are right, thanks !

**Comment by Bronek** [comment]
> I have a few clarifying questions for this PR.
> 
> Although the unit tests are sufficient, I'd like to test the API changes via the command line. How can I enable the API-version 2 in my local `rippled`? I set `beta_rpc_api` to `1` in the config file, but that didn't seem to help.

I do not think it's possible to select API version in command line; @mDuo13 ?

**Comment by intelliot** [comment]
Yes, I think we can always default to the most recent API version for command line. As [stated here](https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md#api-version-1), the command line is intended for ad-hoc usage by humans, not programs or automated scripts. The command line is not meant for use in production code. 

I don't think anyone is using tx_history, not even from the command line. If they are, they are welcome to later open an issue, where we can direct them to use alternative method to get the data they're looking for.

**Review by ckeshava — APPROVED** [review]

**Comment by ximinez** [comment]
@ckeshava Change `apiMaximumSupportedVersion` to 2 and rebuild. The command line uses the maximum _non-beta_ API version.
https://github.com/XRPLF/rippled/blob/1.12.0/src/ripple/net/impl/RPCCall.cpp#L1485

File: src/test/rpc/Handler_test.cpp
Comment (ximinez) [inline src/test/rpc/Handler_test.cpp:42]: Spelling nit: "naive"

File: src/test/rpc/Handler_test.cpp
Comment (ximinez) [inline src/test/rpc/Handler_test.cpp:131]: Since this is testing performance timing and not really correctness, I think it could be changed to a manual test.
```
BEAST_DEFINE_TESTSUITE_MANUAL(Handler, rpc, ripple);
```

**Comment by intelliot** [comment]
note: we intend to [promote](https://github.com/XRPLF/rippled/pull/4568) API version 2 to non-beta after #4733 is merged, and maybe after #4775; and almost certainly before the first 2.0.0-rc

File: src/test/basics/PerfLog_test.cpp
Comment (ximinez) [inline src/test/basics/PerfLog_test.cpp:314]: You're taking this `set` and converting it to a `vector` a couple of times in unit tests. The only other usage of `getHandlerNames()` is in `PerfLog`, where the collection type doesn't matter as long as it's iterable.

I was going to suggest changing the return type back to `vector`, but the overhead to copy the set is unnecessary in the non-test code. What do you think about adding a function to `TestHelpers.h/cpp`, that returns a `vector`? It could even have the same name, since it would be in the `ripple::test` namespace.

File: src/test/rpc/Handler_test.cpp
Comment (ximinez) [inline src/test/rpc/Handler_test.cpp:48]: I don't think the test class shouldn't be in the `ripple::RPC` namespace. `ripple::test` makes more sense.

**Review by ximinez — CHANGES_REQUESTED** [review]
A few minor comments. Overall looks great.

File: src/ripple/rpc/impl/Handler.cpp
Comment (Bronek) [reply src/ripple/rpc/impl/Handler.cpp:215]: Yes, exactly, thank you !

File: src/test/rpc/Handler_test.cpp
Comment (Bronek) [reply src/test/rpc/Handler_test.cpp:131]: Oh great, wan't aware of that macro

**Comment by intelliot** [comment]
@Bronek please comment here to let us know when you think all suggestions have been addressed

File: src/ripple/rpc/handlers/LedgerHandler.h
Comment (Bronek) [reply src/ripple/rpc/handlers/LedgerHandler.h]: Thanks, done

File: src/ripple/rpc/handlers/Version.h
Comment (Bronek) [reply src/ripple/rpc/handlers/Version.h]: Thanks, done.

File: src/ripple/rpc/impl/Handler.cpp
Comment (Bronek) [reply src/ripple/rpc/impl/Handler.cpp]: Yes, that looks better - thanks !

File: src/test/rpc/Handler_test.cpp
Comment (Bronek) [reply src/test/rpc/Handler_test.cpp]: Thanks, fixed

File: src/test/basics/PerfLog_test.cpp
Comment (Bronek) [reply src/test/basics/PerfLog_test.cpp]: Thanks, added a helper function in `TestHelpers.h`

**Comment by Bronek** [comment]
> @Bronek please comment here to let us know when you think all suggestions have been addressed

Yes, all addressed

File: src/test/rpc/Handler_test.cpp
Comment (Bronek) [reply src/test/rpc/Handler_test.cpp:48]: Thanks, moved to `test`

**Comment by Bronek** [comment]
This is ready to merge

**Comment by intelliot** [comment]
awaiting updated review from ximinez (verifying the changes that were made)

**Comment by Bronek** [comment]
> awaiting updated review from ximinez (verifying the changes that were made)

Well in that case I will have one more minor improvement ...

**Review by ximinez — APPROVED** [review]
Looks good!

**Comment by intelliot** [comment]
Suggested commit message:
```
APIv2: remove tx_history and ledger_header (#4759)

Remove `tx_history` and `ledger_header` methods from API version 2.

Update `RPC::Handler` to allow for methods (or method implementations)
to be API version specific. This partially resolves #4727. We can now
store multiple handlers with the same name, as long as they belong to
different (non-overlapping) API versions. This necessarily impacts the
handler lookup algorithm and its complexity; however, there is no
performance loss on x86_64 architecture, and only minimal performance
loss on arm64 (around 10ns). This design change gives us extra
flexibility evolving the API in the future, including other parts of
#4727.

In API version 2, `tx_history` and `ledger_header` are no longer
recognised; if they are called, `rippled` will return error
`unknownCmd`

Resolve #3638

Resolve #3539
```

**Pull request merged**
